Update snapshot summary on rewrite

alex-kattathra-johnson opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
The snapshot summary doesn't reflect the files that are tracked after a rewrite

<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Fixes https://github.com/restic/restic/issues/4902
<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].
-->

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -217,6 +217,31 @@ func filterAndReplaceSnapshot(ctx context.Context, repo restic.Repository, sn *r
 	sn.Original = sn.ID()
 	sn.Tree = &filteredTree
 
+	sn.Summary.TotalFilesProcessed = 0
```
For old snapshots, summary is a nil pointer, such that this would cause a crash.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -217,6 +217,31 @@ func filterAndReplaceSnapshot(ctx context.Context, repo restic.Repository, sn *r
 
+	sn.Summary.TotalFilesProcessed = 0
+	sn.Summary.TotalBytesProcessed = 0
+	processNode := func(_ restic.ID, nodepath string, node *restic.Node, err error) error {
```
Recalculating the snapshot size is rather expensive and only makes sense if `len(rejectByNameFuncs) > 0`.

Iterating through the tree a second time is rather inefficient, instead this recalculation must happen in the `TreeRewriter` in internal/walker/rewriter.go (but only if `DisableNodeCache` is true) where it can be done nearly for free.

alex-kattathra-johnson replied:
By updating snapshot summaries of all snapshots that have a certain path, would we change snapshots we initially didn't mean to touch. I think the recalculating of the size should only be relevant to the specified snapshot. No reference of snapshot gets passed into the `TreeRewriter`.

alex-kattathra-johnson replied:
Would referencing snapshots in `internal/walker/rewriter.go` cause a circular referencing? Maybe we should update in the `RewriteNode` callback like in the latest commit.

alex-kattathra-johnson replied:
I suppose the execution takes place in `internal/walker/rewriter.go`, haha.

MichaelEischer reviewed (changes requested):
Thanks for giving this a try. Recalculating the snapshot size is definitely useful. However, the calculation itself does not belong into `cmd_rewrite.go`, see comments below for more details.

alex-kattathra-johnson commented:
@MichaelEischer : Thanks for the feedback! Updated the PR to make the change to the summary only if it's non-nil and moved the update to the `RewriteNode` callback function.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -136,6 +136,10 @@ func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *resti
 				if selectByName(path) {
 					return node
 				}
+				if sn.Summary != nil && node.Type == "file" {
```
This will break if any directory is excluded.

MichaelEischer replied:
Instead please count all files returned by `return node` and use that to calculate the size from scratch.

MichaelEischer replied:
I'm not yet entirely sure that this is the best place to implement the size calculation. Here it is pretty hard to test it properly. What would be much more testable is SizeCalculationgRewriter that wraps `walker.NewTreeRewriter` and also the passed in `RewriteNode` function to count the remaining snapshot size.

To be a bit more specific I have something like the following in mind (I'm not entirely happy about the `func() SnapshotSize` part):
```go

type SnapshotSize struct {
	FileCount int
	FileSize  int64
}

func NewSizeCountingRewriter(rewriteNode NodeRewriteFunc) (*TreeRewriter, func() SnapshotSize) {
	count := 0
	size := int64(0)
	return NewTreeRewriter(RewriteOpts{
			RewriteNode: func(node *restic.Node, path string) *restic.Node {
				node = rewriteNode(node, path)
				if node != nil {
					// accounting
				}
				return nil
			},
			DisableNodeCache: true,
		}), func() SnapshotSize {
			return SnapshotSize{count, size}
		}
}
```

That would allow using the test infrastructure that already exists in `internal/walker/rewriter_test.go`. With that in place, a single test in `cmd_rewrite.go` would be enough to provide a decent test coverage.

alex-kattathra-johnson replied:
@MichaelEischer : Could you take another look? I ended up creating a function which accepts with snapshot summary as input to be updated in place. And added some tests too.

Commit:
Update snapshot summary on rewrite

Signed-off-by: Alex Johnson <hello@alex-johnson.net>

MichaelEischer commented on internal/walker/rewriter.go:
```diff
@@ -10,8 +10,14 @@ import (
 )
 
 type NodeRewriteFunc func(node *restic.Node, path string) *restic.Node
+type SnapshotSizeRewriteFunc func() SnapshotSize
```
I'd prefer calling this function `QueryRewrittenSizeFunc` or similar. Currently, it sounds like it actually rewrites the snapshot's size, but that isn't the case.

MichaelEischer commented on internal/walker/rewriter.go:
```diff
@@ -52,6 +58,32 @@ func NewTreeRewriter(opts RewriteOpts) *TreeRewriter {
+		RewriteNode: func(node *restic.Node, path string) *restic.Node {
+			node = rewriteNode(node, path)
+			if node != nil {
+				if node.Type == "file" {
```
both `if` can be merged when the function always returns `node`.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -134,20 +134,29 @@ func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *resti
+			return nil
+		}
+
+		rewriter, snapshotSizeRewriter := walker.NewSnapshotSizeRewriter(rewriteNode)
```
please rename `snapshotSizeRewriter` to `querySize`

MichaelEischer commented on internal/walker/rewriter_test.go:
```diff
@@ -303,6 +303,69 @@ func TestRewriter(t *testing.T) {
+
+		ss := snapshotSizeRewriter()
+
+		if ss.FileCount != 2 {
```
Please replace the if-block with `test.Equals(t, uint(2), ss.FileCount, "snapshot file count mismatch")`. (Similarly for `FileSize` below).

MichaelEischer reviewed (commented):
Thanks for the changes so far. I have only a few small nitpicks left, see below.

I'd also like to have a minimal functionality test in `cmd/restic/cmd_rewrite_integration_test.go`, maybe by extending `TestRewriteReplace` to check the snapshot's stats before and after rewriting. I can take care of that if you want.

MichaelEischer reviewed (approved):
LGTM. Thanks!
